net: lwm2m: Add API for asynchronous responses

rlubos opened the pull request:
This PR introduces an API allowing the application to pause/resume message processing (either when generating a response or notification) from the registered resource callbacks. That way, for example if the data requested by the LwM2M engine from the application via read callback is not available, the application can postpone the message processing instead of blocking in the callback. In result, the LwM2M engine can process consecutive requests received while requested data is being prepared, which was not the case if the callback would block.

The overall chain of action looks like this:
 * Application requests to delay message processing with `lwm2m_response_postpone()`. For each delayed reply an additional `struct lwm2m_message` is allocated, where the original request and some additional metadata (resource path where the processing stopped, a pointer to the output context) is stored, as this information is needed when resuming.
 * For READ operation or notifications, we additionally need to store the content output context, so the response/notification can be resumed at correct place in the response packet. For this I've decided to use heap, at least for now. There's just too much variation in the formaters context structures to preallocate those.
 * All postponed messages are stored on a list, in case LwM2M enigne is shut down all get released.
 * When application for example has requested data, or is done with write processing etc., it can resume message processing with `lwm2m_response_resume()`. This marks the message as ready to process and wakes up the engine. The engine will process the request from the beginning, but will take no action for the resources in the request that have already been processed.

I've also added a Generic Sensor object to the LwM2M client sample, demonstrating how this async message processing can be used.

As for testing, so far this has only been manually tested with all content formats and operation types (single resource/multi resource/composite). We'd probably need to add some automated tests for the feature, but I haven't gotten into that yet, and I wanted to open the PR for review before leaving for vacation.  

The feature is marked as experimental for now due to overall complexity. 

Resolves #73224

rlubos committed:
net: lwm2m: Rename processed_req pointer

The LwM2M context kept a pointer to the currently processed request,
however it could also be repurposed to store for example a pointer to
currently processed notification, therefore rename the pointer to
processed_msg to make it more versatile.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: lwm2m: Add helper function to create error response

Move the code responsible for creating error response to a helper
function. Additionally, in case error message generation failed, handle
this gracefully instead of ignoring that fact.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: lwm2m: Add helper function to allocate pending/reply structs

Move the code responsible for allocating pending/reply structs for the
LwM2M message into a helper function so that it can be reused elsewhere.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: lwm2m: Add API for asynchronous responses

Introduce an API allowing the application to pause/resume message
processing (either when generating a response or notification) from the
registered resource callbacks. That way, for example if the data
requested by the LwM2M engine from the application via read callback is
not available, the application can postpone the message processing
instead of blocking in the callback. In result, the LwM2M engine can
process consecutive requests received while requested data is being
prepared, which was not the case if the callback would block.

This commit introduces support for READ requests. Other request types
and notifications support will be introduced in consecutive commits.

As the feature requires to retain the output context for the generated
response message, it's been decided to use heap memory for this case,
given the differences between different content formatters. Therefore
enabling the feature introduces heap dependency.

The feature can be enabled with CONFIG_LWM2M_ASYNC_RESPONSES Kconfig
option and is marked as experimental for now given the overall
complexity of the deferred message processing. In case the option is not
enabled, or the async APIs are not used, the changes should have no
impact on the current behavior..

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: lwm2m: Add support for postponing write operation

This commit introduces support for postponing responses for write
requests.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: lwm2m: Add support for postponing execute operations

This commit introduces support or postponing execute operation requests.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: lwm2m: Add support for postponing notifications

This commit introduces support for postponing sending notifications
when notifications are generated by the LwM2M engine.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
samples: net: lwm2m_client: Demonstrate async operations

Demonstrate how to use asynchronous responses API in a dummy Generic
Sensor object. The object implements delayed read/write/execute
operations.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_message_handling.c:
```diff
@@ -102,12 +102,435 @@ sys_slist_t *lwm2m_engine_obj_list(void);
+		goto out_null;
+	}
+
+	if (sys_slist_find(&client_ctx->postponed_responses, &response->node, NULL)) {
+		goto out_null;
+	}
```
When can this happen?

rlubos committed:
tests: net: lwm2m: Add interop tests for asynchronous operations

Add python tests for asynchronous operations (read, write, exec,
read-composite, write-composite) utilizing all available content
formatters.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos commented:
The branch got pretty old so rebased to make sure everything still works after network symbols overhaul.

@SeppoTakalo Please revist the PR when you have some time.
